lib/galaxy/tools/remote_tool_eval.py

```diff
@@ -60,7 +62,7 @@ def __init__(
         file_sources: ConfiguredFileSources,
     ):
         self.model = Bunch(context=sa_session)
-        self.config = tool_app_config
+        self.config = cast(GalaxyAppConfiguration, tool_app_config)
```

Comment:
It isn't that - adding your extra typing to the app protocol and then lying to mypy here is... counter productive IMO.

Reply:
In that case, let's keep it as is? @nsoranzo if you don't mind, I can revert the commit (and just use 
`# type: ignore[arg-type]`)

Reply:
@jdavcs Sorry for hijacking your PR!

@jmchilton I think we agree that adding typing to a protocol is useful, it's the whole point of a protocol. I'll try to explain why I think that a cast is not a lie but a contract with the type checker.
In unit tests we often mock classes, and if we want to type check tests we need to use workarounds when passing mock objects to functions that expect "real" ones.
I see 3 options, in ascending order of personal preference:
1. Add protocols for each class we need to mock (lots of extra code needed only because of the tests).
2. Add `# type: ignore[assignment]` at every line where the mocked object is passed to a type-annotated function (duplication; not explicit about which type the mocked object simulates; may hide bugs if other parameters of the function call are mis-annotated).
3. Cast the mock object to the "real" type (as here), making explicit to whoever reads the code that we guarantee that for type checking purposes the mock object is equivalent to the "real" one.

Reply:
I have complex feelings but I think I am much more comfortable with the cast in the unitests than the typing that Protocol property as ``GalaxyAppConfiguration``. The code is going to be executed and work or not - so typing is much less useful. Documentation is also less important.

In general, I do agree adding typing to a protocol is useful - obviously. But as a specific issue, I do not think much should be typed as ``GalaxyAppConfiguration``. It is a bad thing IMO. These components should be able to take other kinds of configuration objects (I can point to many places in the code where we setup much more explicit, useful contracts around views on config).

GalaxyAppConfiguration is mostly dynamically generated - so the typing you get is minimal. Even after that it inherits ``HasDynamicProperties`` - to basically tell mypy just let whatever fly.

```
from typing import (
    Any,
    TYPE_CHECKING,
)

if TYPE_CHECKING:

    class HasDynamicProperties:
        def __getattr__(self, property: str) -> Any:
            return object()

else:
    HasDynamicProperties = object
```

So you're introducing a very explicit dependency in that code on a very specific implementation of a God object. You're declaring these components cannot be used in other contexts without really any added typing. You're taking a Galaxy anti-pattern and enshrining it as contract. The best thing you can say about ``config`` is that it is dynamic and you can send a Bunch or something in place of it - and you're taking that away without benefits.

I will stop complaining about casts in unit test code (remind me of the promise if I forget) - but that isn't the root concern here.

Reply:
Thanks for taking the time to detail your reasoning, this makes total sense and I'll restore the `Any` in the protocol for `config` with a comment explaining why it should stay like that.

Reply:
Thank you and thank you for pushing me on it - I made a vague drive by comment based on a strong but unarticulated gut feeling and that wasn't professional or respectful. I should have taken the time to write down my thoughts and be more explicit.

Also thank you both for the comments on unittests and cast - I appreciate all the thoughtful ideas.